### PR TITLE
Create/Drop Statemement binding

### LINKED ISSUE #484: Logical + Physical Operators for Create and Drop
Currently we don't have logical/physical operators for `CREATE` and `DROP` but we do have the plans (i.e `src/include/planner/plannodes/create_*.h` and `src/include/planner/plannodes/drop_*.h`).

Ideally for create/drop operations, the optimizer should take in a logical operator, convert it to a physical operator, and then emit the correct plan (as opposed to special casing in Peloton). To do this, we need logical + physical operators for `CREATE` and `DROP` that contain all the information required by the plan nodes.

### PR DESCRIPTION
**Pull request by mush-zhang**
This pr is an attempt to fix #484 . It also add comparison and hash function for select statement, column definition, index schema, and their related classes. I also fill the missing comparison and hash in some plannodes @mbutrovich (I wonder if those changes are correct). I left some TODOs in the code meant for discussion. 
- [x] create database
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create table
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create index
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create schema
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create trigger
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create function
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] create view
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] drop database
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] drop table
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test
- [x] drop index
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test   
- [x] drop schema
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
  - [x] binding
  - [x] binder test
  - [x] transforming
  - [x] transformer test  
- [x] drop trigger
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test
- [x] drop view
  - [x] logical op
  - [x] logical op test
  - [x] physical op
  - [x] physical op test

### CODE REVIEW

**Comment by mush-zhang** [comment]
Need to do error checking of non-existing table/column reference in binder

File: src/include/catalog/schema.h [hunk 1]
```diff
@@ -293,6 +334,33 @@ class Schema {
...
+    return hash;
+  }
+
+  /**
+   * Perform a comparison of schema
+   * @param rhs other schema to compare against
+   * @return true if the two schema are equal
+   */
+  bool operator==(const Schema &rhs) const {
+    // TODO(Ling): Does column order matter for compare equal?
```
Comment (mbutrovich) [inline src/include/catalog/schema.h]: I believe so.

File: src/include/optimizer/logical_operators.h [hunk 2]
```diff
@@ -982,5 +983,732 @@ class LogicalExportExternalFile : public OperatorNode<LogicalExportExternalFile>
...
+   * @param table_oid OID of the table to be dropped
+   * @return
+   */
+  static Operator Make(catalog::table_oid_t table_oid);
+
+  bool operator==(const BaseOperatorNode &r) override;
+  common::hash_t Hash() const override;
+
+  /**
+   * @return the OID of the database we want to drop
```
Comment (mbutrovich) [inline src/include/optimizer/logical_operators.h]: `return the OID of the table we want to drop`

File: src/include/optimizer/physical_operators.h [hunk 3]
```diff
@@ -1155,5 +1159,761 @@ class Distinct : public OperatorNode<Distinct> {
...
+                       std::unique_ptr<catalog::Schema> table_schema,
+                       common::ManagedPointer<storage::BlockStore> block_store, bool has_primary_key,
+                       planner::PrimaryKeyInfo primary_key, std::vector<planner::ForeignKeyInfo> &&foreign_keys,
+                       std::vector<planner::UniqueInfo> &&con_uniques, std::vector<planner::CheckInfo> &&con_checks);
+
+  bool operator==(const BaseOperatorNode &r) override;
+  common::hash_t Hash() const override;
+
+  /**
+   * @return OID of the namespace to create index on
```
Comment (mbutrovich) [inline src/include/optimizer/physical_operators.h]: `OID of the namespace to create table in`

File: src/include/optimizer/physical_operators.h [hunk 4]
```diff
@@ -1155,5 +1159,761 @@ class Distinct : public OperatorNode<Distinct> {
...
+
+  /**
+   * OID of the trigger to drop
+   */
+  catalog::trigger_oid_t trigger_oid_;
+
+  /**
+   * Whether "IF EXISTS" was used
+   */
+  bool if_exists_;
```
Comment (mbutrovich) [inline src/include/optimizer/physical_operators.h]: Do we need if_exists_ flags in the operators? We should be able to catch this at the Binder level in the future.

File: test/binder/binder_test.cpp [hunk 5]
```diff
@@ -818,4 +819,121 @@ TEST_F(BinderCorrectnessTest, BindDepthTest) {
...
+  BINDER_LOG_DEBUG("Checking create database.");
+
+  std::string create_sql = "CREATE DATABASE C;";
+  auto parse_tree = parser_.BuildParseTree(create_sql);
+  auto statement = parse_tree.GetStatements()[0];
+  EXPECT_NO_THROW(binder_->BindNameToNode(statement, &parse_tree));
+}
+
+// NOLINTNEXTLINE
+TEST_F(BinderCorrectnessTest, CreateTableTest) {
```
Comment (mbutrovich) [inline test/binder/binder_test.cpp]: How do we handle create table with a primary key?

Comment (mush-zhang) [reply test/binder/binder_test.cpp]: I think the binder nor the transformer need to explicitly handle that. 

Primary key info is handled by parser. CreateStatement that corresponds to a create table query has vector of ColumnDefinitions that each embeds a `is_primary_` attribute. I'm letting Binder to not unwrap the columnDefinition until the query is transformed to a physical operator.

Comment (mush-zhang) [reply test/binder/binder_test.cpp]: > How do we handle create table with a primary key?



I think the binder nor the transformer need to explicitly handle that.

Primary key info is handled by parser. CreateStatement that corresponds to a create table query has vector of ColumnDefinitions that each embeds a is_primary_ attribute. I'm letting Binder to not unwrap the columnDefinition until the query is transformed to a physical operator.

File: src/include/optimizer/physical_operators.h (hunk 4 above)
Comment (mush-zhang) [reply src/include/optimizer/physical_operators.h]: > Do we need if_exists_ flags in the operators? We should be able to catch this at the Binder level in the future.

What is standard way for binder to handle this? When Binder detects that the trigger in the "drop trigger if exists" query does not exist, do we create a "no op" operator?

Comment (mbutrovich) [reply src/include/optimizer/physical_operators.h]: Good question, I think we want some way to indicate back to the traffic cop that binding succeeded but execution of the query doesn't need to proceed any further. A noop operator is an interesting idea though for consistency purposes.

File: src/include/optimizer/logical_operators.h [hunk 6]
```diff
@@ -982,5 +983,732 @@ class LogicalExportExternalFile : public OperatorNode<LogicalExportExternalFile>
...
+   * @return trigger when clause
+   */
+  common::ManagedPointer<parser::AbstractExpression> GetTriggerWhen() const {
+    return common::ManagedPointer(trigger_when_);
+  }
+
+  /**
+   * @return trigger type, i.e. information about row, timing, events, access by pg_trigger
+   */
+  int16_t GetTriggerType() const { return trigger_type_; }
```
Comment (apavlo) [inline src/include/optimizer/logical_operators.h]: We should make trigger type an enum. We don't need it for TPC-C so it doesn't need to be fixed now.

Can you write an issue about it and we can fix it after we merge this? Mark it as 'beginner'

Comment (mush-zhang) [reply src/include/optimizer/logical_operators.h]: https://github.com/cmu-db/terrier/issues/655

**Pull request merged**